<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Ralph   University of Oregon   Dept of Data Science">
  <title>Uncovering the time axis of recombining genomes using ARGs</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="./reveal.js/dist/reset.css">
  <link rel="stylesheet" href="./reveal.js/dist/reveal.css">
  <style>
    .reveal .sourceCode {  /* see #7635 */
      overflow: visible;
    }
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    /* The extra [class] is a hack that increases specificity enough to
       override a similar rule in reveal.js */
    ul.task-list[class]{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      font-size: inherit;
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="./reveal.js/dist/theme/simple.css" id="theme">
  <style type="text/css">

  .reveal { font-size: 24px; }

  .reveal h1 { font-size: 1.5em; } 

  .reveal h2 { font-size: 1.2em; } 

  .reveal .slides { text-align: left; }

  .reveal .slides figure { text-align: center; }

  .centered p { text-align: center; }

  /* apply this to the columns div */
  div.vcentered { display: flex; align-items: center; }

  .caption p { font-size: 75%; font-style: italic; color: #0006; }

  .greyed { color: #4444; }

  div.floatright { float: right; text-align: right; }

  .reveal figcaption { display: none; }

  .reveal img { border: none; }

  .reveal section img { border: none; }

  .MathJax_Display { margin: 0; }

  img.finger { vertical-align: text-top; }

  .smaller { font-size: 50%; }
  .smallish { font-size: 80%; }

  .hide { opacity: 0; }

  .red { color: red; }
  .blue { color: blue; }

  .coverup { background: white; }

  </style>

  <div style="display: none">
  \[
  \usepackage{amsmath}
  \usepackage{amssymb}
  \usepackage{xcolor}

  \newcommand{\R}{\mathbb{R}}
  \newcommand{\Q}{\mathbb{Q}}
  \newcommand{\Z}{\mathbb{Z}}
  \newcommand{\N}{\mathbb{N}}
  \newcommand{\C}{\mathbb{C}}

  \renewcommand{\P}{\mathbb{P}}
  \newcommand{\E}{\mathbb{E}}

  \newcommand{\var}{\mathop{\mbox{Var}}}
  \newcommand{\cov}{\mathop{\mbox{cov}}}
  \newcommand{\median}{\mathop{\mbox{median}}}
  %\newcommand{\det}{\mathop{\mbox{det}}}
  \newcommand{\supp}{\mathop{\mbox{supp}}}
  \newcommand{\sgn}{\mathop{\mbox{sgn}}}

  \newcommand{\conv}{\mathop{\mbox{conv}}}
  \newcommand{\deq}{\stackrel{\scriptscriptstyle{d}}{=}}
  \newcommand{\dcv}{\stackrel{\scriptscriptstyle{d}}{\longrightarrow}}

  \]
  </div>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Uncovering the time axis of recombining genomes
using ARGs</h1>
  <p class="author">Peter Ralph <br/> University of Oregon <br/> Dept of
Data Science</p>
  <p class="date">Quantitative and Computational Biology Seminar<br/>
University of Southern California // 5 December 2024</p>
</section>

<section>
<section id="outline" class="title-slide slide level1">
<h1>Outline</h1>
<div class="columns">
<div class="column" style="width:60%;">
<ol type="1">
<li>Ancestral Recombination Graphs (ARGs)</li>
<li>Tree sequences: storage and computation</li>
<li>Variational dating of ARGs</li>
<li>Application: pathogenic variants</li>
</ol>
</div><div class="column" style="width:60%;">
<div class="fragment floatright">
<figure>
<img data-src="figs/tskit_logo.png" style="width:60.0%"
alt="tskit logo" />
<figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<p><a href="https://tskit.dev">tskit.dev</a></p>
</div>
</div>
</div>
<!--
Abstract: The Ancestral Recombination Graph (or, ARG) is a way of representing all ancestral relationships
between a collection of recombining genomes. It can be viewed as a collection of trees along the genome
or as a set of inheritance relationships between modern and ancestral haplotypes. If known, it gives
a more natural and informative way to summarize genetic variation: for instance, by providing ages and
inheritance patterns of the mutations that led to modern variants. In this talk I will summarize recent
progress in estimating whole-genome ARGs, including from Biobank-scale datasets, and describe how this
way of representing data leads to orders-of-magnitude improvements in storage and computation. I will also
describe a (variational) Bayesian method for estimating the times of events in a given ARG, leading to
substantial improvements in dating accuracy. Using the new method on ARGs estimated from 50,000 whole genomes,
I will show that inferred dates are more informative about SNP pathogenicity than allele frequency.
This is one example of how ARGs provide a way to summarize genetic variation without relying on
the more usual problematic and inaccurate division into "populations".
-->
<!--
title-slide-attributes:
    data-background-image: /path/to/title_image.png
    data-background-size: contain
-->
</section>
<section class="slide level2">

<blockquote>
<p>UO is located on the traditional indigenous homeland of the Kalapuya
people. Kalapuya people were dispossessed of their indigenous homeland
by the United States government and forcibly removed. Today, Kalapuya
descendants are primarily citizens of the Confederated Tribes of Grand
Ronde and the Confederated Tribes of Siletz Indians, and continue to
make important contributions to their communities, to the UO, to Oregon,
and to the world.</p>
</blockquote>
</section></section>
<section>
<section id="whats-an-arg" class="title-slide slide level1">
<h1>What’s an ARG?</h1>

</section>
<section id="genomes" class="slide level2">
<h2>Genomes</h2>
<ul>
<li class="fragment">are very big (<span
class="math inline">\(10^7\)</span>–<span
class="math inline">\(10^{12}\)</span> nucleotides)</li>
<li class="fragment">encode the basic <em>mechanisms</em> of life</li>
<li class="fragment">reflect past <em>history</em> and
<em>process</em></li>
</ul>
<div class="fragment">
<figure>
<img data-src="figs/kreitman/kreitman-1983.png"
alt="Genotype matrix from Kreitman 1983" />
<figcaption aria-hidden="true">Genotype matrix from Kreitman
1983</figcaption>
</figure>
</div>
</section>
<section id="meiosis-recombination" class="slide level2">
<h2>Meiosis &amp; Recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<figure>
<img data-src="figs/recombination-cartoon-crop.png"
alt="recombination" />
<figcaption aria-hidden="true">recombination</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<ul>
<li><p>You have two copies of each chromosome, one from each
parent.</p></li>
<li><p>When you make a gamete, the copies <em>recombine</em>, at
Poisson(<span class="math inline">\(\rho\)</span>) locations.</p></li>
<li><p><em>Mutations</em> appear at Poisson(<span
class="math inline">\(\mu\)</span>) locations.</p></li>
</ul>
</div>
</div>
</section>
<section id="what-is-an-arg" class="slide level2">
<h2>What is an ARG?</h2>
<p>For a set of sampled chromosomes, an ARG describes at each position
along the genome the genealogical tree that says how they are
related.</p>
<figure>
<img data-src="figs/example.svg" alt="Trees along a chromosome" />
<figcaption aria-hidden="true">Trees along a chromosome</figcaption>
</figure>
</section>
<section id="what-is-an-arg-1" class="slide level2">
<h2>What is an ARG?</h2>
<div class="columns">
<div class="column" style="width:30%;">
<p>Better, in some ways, is to think of an ARG as a set of relationships
between haplotypes.</p>
<div class="fragment">
<p>An ARG, broadly, is a graph-like structure that tells us how a set of
sampled chromosomes are related to each other, along a recombining
genome.</p>
</div>
</div><div class="column centered">
<p><img data-src="lit/hudson1991.png" style="width:60.0%" /></p>
</div>
</div>
</section></section>
<section>
<section id="arg-inference-a-quick-overview"
class="title-slide slide level1">
<h1>ARG inference: a quick overview</h1>

</section>
<section class="slide level2">

<!-- see lit/notes.md -->
<div class="columns">
<div class="column" style="width:50%;">
<p>ARG/coalescent theory:</p>
<div>
<ul>
<li class="fragment"><a
href="https://www.sciencedirect.com/science/article/abs/pii/0040580983900138">Hudson
1983, 1991</a></li>
<li class="fragment"><a
href="https://doi.org/https:%2f%2fdoi.org%2f10.1006%2ftpbi.1994.1023">Griffiths
&amp; Tavaré 1994</a></li>
<li class="fragment"><a
href="https://www.bibsonomy.org/bibtex/2d60d21cbcdf8d4328d26446b0ea3bee8/peter.ralph">Griffiths
&amp; Marjoram 1997</a></li>
</ul>
</div>
<div class="fragment">
<p>Parsimony: dozens of samples, dozens of SNPs:</p>
</div>
<div>
<ul>
<li class="fragment"><a
href="https://doi.org/10.1093/genetics/111.1.147">Hudson &amp; Kaplan
1985</a></li>
<li class="fragment"><a
href="https://www.sciencedirect.com/science/article/abs/pii/002555649090123G">Hein
1990</a></li>
<li class="fragment"><a
href="https://dx.doi.org/10.1093/genetics/163.1.375">Myers &amp;
Griffiths 2003</a></li>
<li class="fragment"><a
href="https://doi.org/10.1089/cmb.2005.12.147">Song &amp; Hein
2005</a></li>
<li class="fragment"><a
href="https://doi.org/10.1093/bioinformatics/bti1033">Song, Wu &amp;
Gusfeld 2005</a></li>
<li class="fragment"><a
href="https://pubmed.ncbi.nlm.nih.gov/17033967/">Minichello &amp; Durbin
2006</a></li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p>Full likelihood: integrating over possible ARGs; dozens of samples, a
few dozen SNPs:</p>
</div>
<div>
<ul>
<li class="fragment"><a
href="https://www.liebertpub.com/doi/abs/10.1089/cmb.1996.3.479">Griffiths
&amp; Marjoram 1996</a></li>
<li class="fragment"><a
href="https://academic.oup.com/genetics/article/156/3/1393/6051663">Kuhner,
Yamato &amp; Felsenstein 2000</a></li>
<li class="fragment"><a
href="https://academic.oup.com/genetics/article/154/2/931/6047972">Nielsen
2000</a></li>
<li class="fragment"><a
href="https://academic.oup.com/genetics/article/159/3/1299/6049673">Fearnhead
&amp; Donnelly 2001</a></li>
<li class="fragment"><a
href="https://doi.org/https:%2f%2fdoi.org%2f10.1111%2f1467-9868.00254">Stephens
and Donnelly 2002</a></li>
<li class="fragment">ARGweaver <a
href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004342">(Rasmussen
et al 2014)</a> MCMC; dozens of samples for <em>megabases</em></li>
</ul>
</div>
<div class="fragment">
<p>“Modern” methods (whole chromosomes!):</p>
</div>
<div>
<ul>
<li class="fragment">Relate <a
href="https://doi.org/10.1038%2fs41588-019-0484-x">(Speidel et al
2019)</a> local tree building + reconciliation; <span
class="math inline">\(n \sim 10^3\)</span></li>
<li class="fragment">tsinfer <a
href="https://doi.org/10.1038%2fs41588-019-0483-y">(Kelleher et al
2019)</a> haplotype matching; <span class="math inline">\(n \sim
10^5\)</span></li>
<li class="fragment">ARG-needle <a
href="https://doi.org/10.1038%2fs41588-023-01379-x">(Zhang et al
2023)</a> haplotype matching + <a
href="https://palamaralab.github.io/software/asmc/">ASMC</a>; <span
class="math inline">\(n \sim 10^5\)</span></li>
<li class="fragment">Singer <a
href="https://doi.org/10.1101%2f2024.03.16.585351">(Deng, Nielsen &amp;
Song 2024)</a> MCMC with full likelihood; <span class="math inline">\(n
\sim 10^2\)</span></li>
</ul>
</div>
</div>
</div>
</section>
<section class="slide level2">

<div class="r-stack centered">
<p><img data-src="lit/hudson1991.png" style="width:50.0%" /></p>
<div class="fragment">
<p><img data-src="lit/griffiths-marjoram-1997.png"
style="width:50.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="lit/kuhner-yamato-felsenstein-2000.png"
style="width:60.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="lit/fearnhead-donnelly-2001.png"
style="width:50.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="lit/song-wu-gusfield-2005.png"
style="width:80.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="lit/song-hein-2005.png" style="width:60.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="lit/rasmussen2014.png" style="width:80.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="lit/kelleher2019.png" style="width:70.0%" /></p>
</div>
</div>
</section>
<section class="slide level2">

<p><img data-src="figs/kreitman/inference.png" /></p>
<div class="caption">
<p>from <a
href="https://academic.oup.com/genetics/article/228/1/iyae100/7714980">Wong
et al 2024, <em>Genetics</em></a></p>
</div>
<!-- Tree sequences: storage and computation -->
</section></section>
<section>
<section id="storage-and-computation" class="title-slide slide level1">
<h1>Storage and computation</h1>
<div class="floatright">
<figure>
<img data-src="figs/tskit_logo.png" style="width:40.0%"
alt="tskit logo" />
<figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
</div>
</section>
<section class="slide level2">

<p>One way to view an ARG is as a sequence of genealogical trees (with
many nodes shared between adjacent trees).</p>
<div class="fragment">
<figure>
<img data-src="figs/example.svg" alt="Trees along a chromosome" />
<figcaption aria-hidden="true">Trees along a chromosome</figcaption>
</figure>
</div>
</section>
<section class="slide level2">

<p>The <strong>succinct tree sequence</strong></p>
<div class="floatright">
<p>is a way to succinctly describe this, er, sequence of trees</p>
<p><em>and</em> the resulting genome sequences.</p>
<div class="caption">
<p><a
href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher,
Etheridge, &amp; McVean</a></p>
</div>
</div>
<div class="fragment">
<div class="columns" style="clear: both;">
<div class="column" style="width:50%;">
<figure>
<img data-src="figs/tskit_logo.png" style="width:80.0%"
alt="tskit logo" />
<figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<div class="floatright">
<figure>
<img data-src="figs/jerome.jpeg" style="width:50.0%"
alt="jerome kelleher" />
<figcaption aria-hidden="true">jerome kelleher</figcaption>
</figure>
<div class="caption">
<p>jerome kelleher</p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="example-three-samples-two-trees-two-variant-sites"
class="slide level2">
<h2>Example: three samples; two trees; two variant sites</h2>
<figure>
<img data-src="figs/example_tree_sequence.png"
alt="Example tree sequence" />
<figcaption aria-hidden="true">Example tree sequence</figcaption>
</figure>
</section>
<section id="nodes-and-edges" class="slide level2">
<h2>Nodes and edges</h2>
<dl>
<dt>Edges</dt>
<dd>
<p>Who inherits from who.</p>
<p>Records: interval (left, right); parent node; child node.</p>
</dd>
<dt>Nodes</dt>
<dd>
<p>The ancestors those happen in.</p>
<p>Records: time ago (of birth); ID (implicit).</p>
</dd>
</dl>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.0.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.1.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.2.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.3.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.4.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.5.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.6.png"
alt="Building a tree sequence" />
<figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section id="sites-and-mutations" class="slide level2">
<h2>Sites and mutations</h2>
<dl>
<dt>Mutations</dt>
<dd>
<p>When state changes along the tree.</p>
<p>Records: site it occured at; node it occurred in; derived state.</p>
</dd>
<dt>Sites</dt>
<dd>
<p>Where mutations fall on the genome.</p>
<p>Records: genomic position; ancestral (root) state; ID (implicit).</p>
</dd>
</dl>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.0.png"
alt="Adding mutations" />
<figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.1.png"
alt="Adding mutations" />
<figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.2.png"
alt="Adding mutations" />
<figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.3.png"
alt="Adding mutations" />
<figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.4.png"
alt="Adding mutations" />
<figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<p><strong>The result:</strong> an encoding of the genomes <em>and</em>
all the genealogical trees.</p>
<div class="centered">
<figure>
<img data-src="figs/example_tree_sequence.png"
alt="Example tree sequence" />
<figcaption aria-hidden="true">Example tree sequence</figcaption>
</figure>
</div>
<!-- # How's it work? -->
</section>
<section id="file-sizes" class="slide level2">
<h2>File sizes</h2>
<div class="centered">
<figure>
<img data-src="figs/tsinfer_sizes.png" style="width:90.0%"
alt="file sizes" />
<figcaption aria-hidden="true">file sizes</figcaption>
</figure>
</div>
<div class="caption">
<p>100Mb chromosomes; from <a
href="https://www.nature.com/articles/s41588-019-0483-y">Kelleher et al
2018, <em>Inferring whole-genome histories in large population
datasets</em></a>, Nature Genetics</p>
</div>
<!-- Estimated sizes of files required to store the genetic variation data for a
simulated human-like chromosome (100 megabases) for up to 10 billion haploid
(5 billion diploid) samples. Simulations were run for 10 1 up to 10 7 haplotypes
using msprime [Kelleher et al., 2016], and the sizes of the resulting files plotted
(points). -->
</section>
<section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_genotypes.png"
alt="genotypes" />
<figcaption aria-hidden="true">genotypes</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_genotype_and_tree.png"
alt="genotypes and a tree" />
<figcaption aria-hidden="true">genotypes and a tree</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_next_genotype_and_tree.png"
alt="genotypes and the next tree" />
<figcaption aria-hidden="true">genotypes and the next tree</figcaption>
</figure>
</section>
<section id="for-n-samples-genotyped-at-m-sites" class="slide level2">
<h2>For <span class="math inline">\(N\)</span> samples genotyped at
<span class="math inline">\(M\)</span> sites</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><em>Genotype matrix</em>:</p>
<p><span class="math inline">\(N \times M\)</span> things.</p>
</div><div class="column" style="width:50%;">
<p><em>Tree sequence:</em></p>
<ul>
<li><span class="math inline">\(2N-2\)</span> edges for the first
tree</li>
<li><span class="math inline">\(\sim 4\)</span> edges per each of <span
class="math inline">\(T\)</span> trees</li>
<li><span class="math inline">\(M\)</span> mutations</li>
</ul>
<p><span class="math inline">\(O(N + T + M)\)</span> things</p>
</div>
</div>
<figure>
<img data-src="figs/ts_ex/tree_sequence_genotype_and_tree.png"
style="width:60.0%" alt="genotypes and a tree" />
<figcaption aria-hidden="true">genotypes and a tree</figcaption>
</figure>
<!-- # Summarizing genomes and genealogies -->
</section>
<section id="fast-genotype-statistics" class="slide level2">
<h2>Fast genotype statistics</h2>
<div class="centered">
<figure>
<img data-src="figs/treestats/benchmarks_without_copy_longer_genome.png"
style="width:70.0%" alt="efficiency of treestat computation" />
<figcaption aria-hidden="true">efficiency of treestat
computation</figcaption>
</figure>
</div>
<div class="caption">
<p>from <a
href="https://academic.oup.com/genetics/article/215/3/779/5930459">R.,
Thornton and Kelleher 2019, <em>Efficiently summarizing relationships in
large samples</em></a>, Genetics</p>
</div>
</section>
<section id="summaries-of-genotypes-and-genealogies"
class="slide level2">
<h2>Summaries of genotypes and genealogies</h2>
<div class="columns">
<div class="column" style="width:47%;">
<p><em>Genotypes:</em></p>
<ol type="1">
<li>For each site,</li>
<li>look at who has which alleles,</li>
<li>and add a <em>summary</em> of these values to our running
total.</li>
</ol>
<p><em>Example:</em> genetic distance counts how many mutations differ
between two sequences.</p>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47%;">
<figure>
<img data-src="figs/ts_ex/tree_sequence_site.png" alt="site stats" />
<figcaption aria-hidden="true">site stats</figcaption>
</figure>
</div>
</div>
</section>
<section id="summaries-of-genotypes-and-genealogies-1"
class="slide level2">
<h2>Summaries of genotypes and genealogies</h2>
<div class="columns">
<div class="column" style="width:47%;">
<p><em>Trees:</em></p>
<ol type="1">
<li>For each branch,</li>
<li>look at who would inherit mutations on that branch,</li>
<li>and add the <em>expected contribution</em> to the running
total.</li>
</ol>
<p><em>Example:</em> the mean time to most recent common ancestor
between two sequences.</p>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47%;">
<figure>
<img data-src="figs/ts_ex/tree_sequence_branch.png"
alt="branch stats" />
<figcaption aria-hidden="true">branch stats</figcaption>
</figure>
</div>
</div>
</section>
<section id="section" class="slide level2">
<h2></h2>
<p>Given</p>
<ol type="1">
<li>a <em>weight</em> <span class="math inline">\(w_i \in
\mathbb{R}^n\)</span> for each <em>sample node</em>, and</li>
<li>a <em>summary function</em> <span class="math inline">\(f :
\mathbb{R}^n \to \mathbb{R}\)</span>,</li>
</ol>
<div class="fragment">
<p>the <strong>Site</strong> statistic <span
class="math display">\[\begin{equation}
\text{Site}(f,w) = \sum_{i: \text{sites}} \sum_{a: \text{alleles}_i}
f(w_a)
\end{equation}\]</span> is the total summarized weights of all
mutations,</p>
</div>
<div class="fragment">
<p>and the <strong>Branch</strong> statistic <span
class="math display">\[\begin{equation}
\text{Branch}(f,w) = \sum_{T: \text{trees}} s_T \sum_{b:
\text{branches}_T} f(w_b) \ell_b
\end{equation}\]</span> is the <em>expected value</em> of <span
class="math inline">\(\text{Site}(f,w)\)</span> under Poisson(1)
mutation, given the trees.</p>
</div>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<p>With genealogies <em>fixed</em>, and averaging only over
<em>mutations</em> with rate <span class="math inline">\(\mu\)</span>,
<span class="math display">\[\begin{equation}
    \text{Branch}(f, w) = \frac{1}{\mu} \mathbb{E}\left[ \text{Site}(f,
w) \vert T \right] .
\end{equation}\]</span></p>
<div class="fragment">
<p>Dealing directly with genealogies removes the layer of noise due to
mutation: <span class="math display">\[\begin{equation}
    \frac{1}{\mu^2} \text{var}\left[\text{Site}(f,w)\right]
    =
    \text{var}\left[\text{Branch}(f,w)\right]
    +
    \frac{1}{n}
    \mathbb{E}\left[\text{Branch}(f^2,w)\right]
\end{equation}\]</span></p>
</div>
<div class="fragment">
<p>and might produce <em>unbiased</em> estimates from ascertained
genotype data.</p>
<div class="caption greyed floatright">
<p>also see R., TPB, 2019</p>
</div>
</div>
</section>
<section id="section-2" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<figure>
<img
data-src="figs/treestats/relate_chr20_site_div_branch.1000000.diversity.png"
style="width:100.0%" alt="duality in 1000G data" />
<figcaption aria-hidden="true">duality in 1000G data</figcaption>
</figure>
</div><div class="column" style="width:40%;">
<p>Duality, on 1000 Genomes data? Not quite…</p>
<ul>
<li>variation in mutation rate?</li>
<li>biased gene conversion?</li>
<li>selection?</li>
<li>inference artifacts?</li>
</ul>
<div class="caption">
<p><em>Tree sequence from <a
href="https://www.nature.com/articles/s41588-019-0484-x">Speidel et al
2019</a>.</em></p>
</div>
</div>
</div>
</section></section>
<section>
<section id="the-tree-sequence-as-a-sparse-matrix"
class="title-slide slide level1">
<h1>The tree sequence as a sparse matrix</h1>
<p><em>Problem:</em> given a <span class="math inline">\(N \times
M\)</span> genotype matrix <span class="math inline">\(G\)</span> (of
0/1s) and a <span class="math inline">\(M\)</span>-vector <span
class="math inline">\(v\)</span>, compute <span
class="math inline">\(Gv\)</span>.</p>
<div class="fragment">
<p>If <span class="math inline">\(v_j\)</span> is the additive effect of
SNP <span class="math inline">\(j\)</span>, then <span
class="math inline">\((Gv)_i\)</span> is the genetic value of chromosome
<span class="math inline">\(i\)</span>.</p>
</div>
</section>
<section id="matrix-multiplication" class="slide level2">
<h2>Matrix multiplication</h2>
<p><em>Problem:</em> compute <span class="math inline">\(G1\)</span>,
the number of mutations that each genome carries.</p>
<div class="r-stack">
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_00.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_01.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_02.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_03.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_04.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_05.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_06.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_07.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_08.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_09.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_10.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/tree_sequence/summing_down_11.png" /></p>
</div>
</div>
</section>
<section class="slide level2">

<div class="columns">
<div class="column">
<p><em>Naive complexity:</em> <span
class="math inline">\(O(NM)\)</span></p>
<p><em>Tree sequence:</em> <span class="math inline">\(O(N + M +
T)\)</span></p>
<div class="fragment">
<p>Similar: relatedness matrix-vector product (used in GWAS, genomic
prediction, etc).</p>
</div>
</div><div class="column centered">
<p><img data-src="figs/tree_sequence/summing_down_10.png" /></p>
</div>
</div>
</section></section>
<section>
<section id="application-pathogenic-variants"
class="title-slide slide level1">
<h1>Application: pathogenic variants</h1>

</section>
<section id="section-3" class="slide level2"
data-background-image="figs/gel-rare-diseases.png"
data-background-position="center" data-background-size="100%">
<h2 data-background-image="figs/gel-rare-diseases.png"
data-background-position="center" data-background-size="100%"></h2>
</section>
<section id="the-uk-100000-genomes-project" class="slide level2">
<h2>the <a
href="https://www.genomicsengland.co.uk/initiatives/100000-genomes-project">UK
100,000 Genomes project</a></h2>
<div class="columns">
<div class="column" style="width:65%;">
<p><img data-src="figs/folks/Sam-Tallman.jpg" style="width:30.0%"
alt="Sam Tallman" /> <img data-src="figs/folks/yan_wong.jpg"
style="width:30.0%" alt="Yan Wong" /> <img
data-src="figs/folks/ben_jeffery.jpg" style="width:30.0%"
alt="Ben Jeffery" /> <img data-src="figs/folks/jerome_kelleher.jpg"
style="width:30.0%" alt="Jerome Kelleher" /> <img
data-src="figs/folks/duncan-mr.jpeg" style="width:30.0%"
alt="Duncan Mbuli-Robertson" /> <img data-src="figs/nate-pope.png"
style="width:30.0%" alt="Nate Pope" /></p>
</div><div class="column" style="width:35%;">
<figure>
<img data-src="figs/tskit_logo.png" style="width:60.0%"
alt="tskit logo" />
<figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<p>Sam Tallman</p>
<p>Yan Wong</p>
<p>Ben Jeffery</p>
<p>Jerome Kelleher</p>
<p>Duncan Mbuli-Robertson</p>
<p>Nate Pope</p>
<div class="fragment">
<p><em>slides in part due to Sam Tallman</em></p>
</div>
</div>
</div>
</section>
<section id="the-100000-genomes-project" class="slide level2">
<h2>The 100,000 Genomes Project</h2>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li><p>UK initiative to sequence 100,000 genomes from 85,000 NHS
patients and family members affected by rare disease or cancer.</p></li>
<li><p>71,800 participants (probands and family members) across over 190
rare diseases, none with genetic diagnoses prior to
recruitment.</p></li>
<li><p>Since 2018, over 6,000 genetic diagnoses</p></li>
</ul>
</div><div class="column" style="width:40%;">
<p><img data-src="figs/gel-logo.png" style="width:90.0%" /></p>
</div>
</div>
</section>
<section id="motivation-penetrant-pathogenic-variants-usually-rare"
class="slide level2">
<h2>Motivation: penetrant, pathogenic variants usually rare</h2>
<div class="centered">
<p><img data-src="figs/sam/variant-classification.jpg" /></p>
</div>
<div class="caption">
<p><a href="https://www.nature.com/articles/gim201737">Nykamp et
al. 2017</a></p>
</div>
</section>
<section id="but-datasets-are-notoriously-unbalanced"
class="slide level2">
<h2>But: datasets are notoriously unbalanced</h2>
<div class="columns">
<div class="columns" width="50%">
<div class="fragment">
<p><img data-src="figs/genome-study-representation.png"
style="width:80.0%" /></p>
<div class="caption">
<p>Sam Tallman</p>
</div>
</div>
</div><div class="columns" width="50%">
<div class="fragment">
<p><img data-src="figs/manrai_2016_nejm.jpg" style="width:100.0%" /></p>
<div class="caption">
<p><a href="https://www.nejm.org/doi/pdf/10.1056/NEJMsa1507092">Manrai
et al 2016</a>, <em>Genetic Misdiagnoses and the Potential for Health
Disparities</em></p>
</div>
</div>
</div>
</div>
</section>
<section id="what-about-age" class="slide level2"
data-background-image="figs/folks/yan_wong.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">
<h2 data-background-image="figs/folks/yan_wong.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">What about age?</h2>
<div class="r-stack">
<div class="fragment">
<p><img data-src="figs/sampling_sim_ages_l1.png"
style="width:90.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/sampling_sim_ages_l1_l2.png"
style="width:90.0%" /></p>
</div>
</div>
</section></section>
<section>
<section id="variational-dating-tsdate-0.2"
class="title-slide slide level1">
<h1>Variational dating: tsdate 0.2</h1>
<div class="centered">
<p><img data-src="figs/tsdate_logo.png" style="width:30.0%" /></p>
</div>
<div class="floatright">
<figure>
<img data-src="figs/nate-pope.png" style="width:50.0%" alt="tsNate" />
<figcaption aria-hidden="true">tsNate</figcaption>
</figure>
</div>
<div class="caption floatright">
<p>Nate Pope</p>
</div>
</section>
<section id="dating-an-arg" class="slide level2">
<h2>Dating an ARG</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><em>Given</em> an ARG with:</p>
<ul>
<li>nodes <span class="math inline">\(\mathcal{N}\)</span></li>
<li>edges <span class="math inline">\(\mathcal{E} \subset \{i \to j : i,
j \in \mathcal{N}\}\)</span></li>
<li>mutation counts per edge <span class="math inline">\(\{y_{ij} : ij
\in \mathcal{E}\}\)</span></li>
<li><em>sample</em> nodes with known time</li>
</ul>
<p><em>Infer:</em></p>
<ul>
<li>times <span class="math inline">\(\{t_i\}\)</span> of remaining
nodes</li>
<li>satisfying the constraints <span class="math inline">\(\{t_i &lt;
t_j : i \to j \in \mathcal{E}\}\)</span></li>
</ul>
</div><div class="column" style="width:50%;">
<div class="r-stack">
<p><img data-src="figs/dating_problem/dating_problem_00.png" /></p>
<div class="fragment">
<p><img data-src="figs/dating_problem/dating_problem_01.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/dating_problem/dating_problem_02.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/dating_problem/dating_problem_03.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/dating_problem/dating_problem_04.png" /></p>
</div>
</div>
</div>
</div>
</section>
<section id="dating-an-arg-1" class="slide level2">
<h2>Dating an ARG</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><em>Goal:</em> find <span class="math inline">\(\{t_i\}\)</span></p>
<p><em>The model:</em> with mutation rate <span
class="math inline">\(\mu\)</span>, edge span <span
class="math inline">\(s_{ij}\)</span>: <span
class="math display">\[y_{ij} \sim \text{Poisson}(\mu s_{ij} (t_j - t_i)
)\]</span></p>
<div class="fragment">
<p><em>The MLE:</em> minimize <span
class="math display">\[\begin{aligned}&amp;\sum_{ij \in \mathcal{E}} \mu
s_{ij} (t_j - t_j) \\&amp;\qquad{}- y_{ij} \log\left(\mu s_{ij} (t_j -
t_j)\right)\end{aligned}\]</span> subject to <span
class="math inline">\(\{t_i &lt; t_j : i \to j \in
\mathcal{E}\}\)</span></p>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/dating_problem/dating_problem_04.png" /></p>
</div>
</div>
</section>
<section id="but-what-about-uncertainty" class="slide level2">
<h2>But, what about uncertainty?</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><em>New goal</em>: infer <span class="math inline">\(t =
\{t_i\}\)</span>, but Bayesian</p>
<div>
<ul>
<li class="fragment"><p>Given a prior <span
class="math inline">\(p(t)\)</span>,</p></li>
<li class="fragment"><p>the per-edge likelihood is <span
class="math display">\[p(y_{ij}|t_i-t_j) \propto (t_j - t_j)^{y_{ij}}
e^{-\mu s_{ij} (t_i - t_j)}\]</span></p></li>
<li class="fragment"><p>the full likelihood is <span
class="math display">\[p(t,y) = p(t) \prod_{ij} p(y_{ij}|t_i -
t_j)\]</span></p></li>
<li class="fragment"><p>and so the marginal posterior on <span
class="math inline">\(t_i\)</span> is <span
class="math display">\[p(t_i|y) = \frac{\int p(t,y) dt_{\setminus
i}}{\int p(t,y) dt}\]</span></p></li>
<li class="fragment"><p>but, those integrals are rather
difficult.</p></li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/dating_problem/dating_problem_05.png"
style="width:120.0%" /></p>
</div>
</div>
</div>
</section>
<section id="variational-approximation" class="slide level2">
<h2>Variational approximation</h2>
<div class="columns">
<div class="column" style="width:70%;">
<ul>
<li><p>Approximate posterior marginals by Gamma distribution</p></li>
<li><p>Fit by matching moments</p></li>
<li><p>Result (hopefully) has exact marginal moments but ignores
dependence in posterior</p></li>
</ul>
</div><div class="column" style="width:70%;">
<div class="r-stack">
<div class="fragment">
<p><img data-src="figs/nate/plot_like.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/nate/plot_exact.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/nate/plot_both.png" /></p>
</div>
</div>
</div>
</div>
<!--
## Expectation propagation

factor variational posterior
into edge terms (like true posterior)

then do iterative refinement of edge contributions

that reduces to adjusting moments for $i$ and $j$
thanks to exponential family magic

and some serious computational magic
-->
</section>
<section id="some-assembly-required" class="slide level2">
<h2>Some assembly required</h2>
<div class="columns">
<div class="column" style="width:70%;">
<ul>
<li>Fit by expectation propagation,</li>
<li>requiring efficient stable computation of <span
class="math inline">\({}_2F_1(a,b;c;z)\)</span>.</li>
</ul>
<p>Additional magic:</p>
<ul>
<li>ambiguous singleton phasing</li>
<li>better prior on root ages</li>
<li>“recalibration” of times to enforce molecular clock</li>
</ul>
</div><div class="column" style="width:70%;">
<p><img data-src="figs/nate-pope-hat.png" /></p>
</div>
</div>
</section>
<section id="validation-simulation" class="slide level2"
data-background-image="figs/folks/yan_wong.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">
<h2 data-background-image="figs/folks/yan_wong.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">Validation: simulation</h2>
<div class="r-stack">
<div class="fragment">
<p><img data-src="figs/sampling_sim_sub_l1_l2.png"
style="width:70.0%" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/sampling_sim_sub_l1_l2_l3.png"
style="width:70.0%" /></p>
</div>
</div>
</section>
<section id="validation-real-data" class="slide level2"
data-background-image="figs/folks/yan_wong.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">
<h2 data-background-image="figs/folks/yan_wong.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">Validation: real data</h2>
<p><img data-src="figs/validation_sub.png" /></p>
</section></section>
<section>
<section id="back-to-the-data" class="title-slide slide level1">
<h1>Back to the data</h1>

</section>
<section id="section-4" class="slide level2"
data-background-image="figs/folks/ben_jeffery.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%">
<h2 data-background-image="figs/folks/ben_jeffery.jpg"
data-background-position="bottom 50px right 50px"
data-background-size="10%"></h2>
<p>We built the world’s largest whole-chromosome tree sequence from the
phased 100,000 Genomes Project (aggV2) dataset.</p>
<p>5,680,570 hypothetical ancestors along chromosome 17 (as an example).
~870Mb.</p>
<p><img data-src="figs/sam/chr17age_vs_freq.png" /></p>
</section>
<section id="adac" class="slide level2">
<h2>ADAC</h2>
<p>Age Divergence at Allele Count (ADAC) quantifies (as an odds ratio)
the probability that a given mutation is older than expected relative to
some reference set of mutations at the same frequency in the data</p>
<div class="r-stack">
<div class="fragment">
<p><img data-src="figs/sam/adac-defn.png" /></p>
</div>
<div class="fragment">
<p><img data-src="figs/sam/adac-defn2.png" /></p>
</div>
</div>
</section>
<section id="deleterious-variants-are-enriched-in-recent-times"
class="slide level2">
<h2>Deleterious variants are enriched in recent times</h2>
<p><img data-src="figs/sam/missense_enrichment.png" /></p>
</section>
<section class="slide level2">

<p><img data-src="figs/sam/enrichment3.png" /></p>
</section>
<section class="slide level2">

<p><img data-src="figs/sam/example-indiv.png" /></p>
</section>
<section id="conclusions" class="slide level2">
<h2>Conclusions</h2>
<ul>
<li><p>Mutations may be rare because they are recent or just because the
individual’s ancestry is not well-represented in the data.</p></li>
<li><p>ARGs capture the complex and continuous nature of human genetic
ancestry.</p></li>
<li><p>Mutations’ age estimation reflects this uncertainty (at single
locus resolution).</p></li>
<li><p>ARG inference and variant dating methods that work well with
large, heterogeneous, and unbalanced biobanks can provide valuable
insights into genetic variation without the need for discrete
categorizations.</p></li>
</ul>
</section></section>
<section>
<section id="wrap-up" class="title-slide slide level1">
<h1>Wrap-up</h1>

</section>
<section id="software-development-goals" class="slide level2">
<h2>Software development goals</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>open</li>
<li>welcoming and supportive</li>
<li>reproducible and well-tested</li>
<li>backwards compatible</li>
<li>well-documented</li>
<li>capacity building</li>
</ul>
<div class="centered">
<figure>
<img data-src="figs/popsim.png" style="width:35.0%" alt="popsim logo" />
<figcaption aria-hidden="true">popsim logo</figcaption>
</figure>
</div>
<p><a
href="https://popsim-consortium.github.io/stdpopsim-docs/stable/index.html">PopSim
Consortium</a></p>
</div><div class="column" style="width:50%;">
<div class="centered">
<figure>
<img data-src="figs/tskit_logo.png" style="width:60.0%"
alt="tskit logo" />
<figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<p><a href="https://tskit.dev">tskit.dev</a></p>
<figure>
<img data-src="figs/slim_logo.png" style="width:80.0%"
alt="SLiM logo" />
<figcaption aria-hidden="true">SLiM logo</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="thanks" class="slide level2">
<h2>Thanks!</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div style="font-size: 85%;">
<ul>
<li>Andy Kern</li>
<li>Nate Pope</li>
<li>Victoria Caudill</li>
<li>Murillo Rodrigues</li>
<li>Gilia Patterson</li>
<li>Chris Smith</li>
<li>Thomas Forest</li>
<li>Jiseon Min</li>
<li>Clara Rehmann</li>
<li>Anastasia Teterina</li>
<li>Angel Rivera-Colon <!--
- Bruce Edelman
- Matt Lukac
- Saurabh Belsare
- Gabby Coffing
- Jeff Adrion
- CJ Battey
- Jared Galloway
--></li>
<li>the rest of <a href="https://kr-colab.github.io/people">the
Co-Lab</a></li>
</ul>
<p>Funding:</p>
<ul>
<li>NIH NIGMS</li>
<li>NSF DBI</li>
</ul>
</div>
<p><img src="figs/KernRalph_5x5.png" alt="KR-colab logo" style="width: 50%; margin: 0px;"/></p>
</div><div class="column" style="width:60%;">
<div class="columns">
<div class="column" style="width:30%;">
<div style="font-size: 85%;">
<ul>
<li>Jerome Kelleher</li>
<li>Ben Haller</li>
<li>Yan Wong</li>
<li>Ben Jeffery</li>
<li>Sam Tallman</li>
<li>Duncan Mbuli-Robertson</li>
<li>Hanbin Lee</li>
<li>Gregor Gorjanc</li>
<li>Elsie Chevy</li>
<li>Madeline Chase</li>
<li>Sean Stankowski</li>
<li>Matt Streisfeld <!--
- Georgia Tsambos
- Jaime Ashander
- Jared Galloway
- Gideon Bradburd
- Bill Cresko
- Alison Etheridge
- Evan McCartney-Melstad
- Brad Shaffer
--></li>
</ul>
</div>
</div><div class="column" style="width:30%;">
<div class="centered" style="margin-top: -60px;">
<p><img src="figs/tskit_logo.png" alt="tskit logo" style="width: 60%; margin: 10px;"/>
<img src="figs/slim_logo.png" alt="SLiM logo" style="width: 90%; margin: 10px;"/></p>
</div>
</div>
</div>
<div style="margin-top: -30px;">
<p><img data-src="figs/colab.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section>
<section id="section-5" class="slide level2"
data-background-image="figs/guillemots_thanks.png"
data-background-position="center" data-background-size="50%">
<h2 data-background-image="figs/guillemots_thanks.png"
data-background-position="center" data-background-size="50%"></h2>
</section></section>
    </div>
  </div>

  <script src="./reveal.js/dist/reveal.js"></script>

  <!-- reveal.js plugins -->
  <script src="./reveal.js/plugin/notes/notes.js"></script>
  <script src="./reveal.js/plugin/search/search.js"></script>
  <script src="./reveal.js/plugin/zoom/zoom.js"></script>
  <script src="./reveal.js/plugin/math/math.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
        // Display controls in the bottom right corner
        controls: true,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: true,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'bottom-right',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: true,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: false,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: true,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'default',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: true,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'fade',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        math: {
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [
          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    </body>
</html>
